pal/src/sync/cclock.cpp

```diff
@@ -0,0 +1,78 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <sched.h>
+#include <pthread.h>
+#include <assert_only.h>
+#include <cclock.hpp>
+#include <errno.h>
+// do not include PAL
+
+void CCLock::Reset(bool shouldTrackThreadId)
+{
+    Assert(sizeof(pthread_mutex_t) <= sizeof(this->mutexPtr));
+
+    int err;
+    pthread_mutexattr_t mtconf;
+    if (shouldTrackThreadId)
+    {
+        err = pthread_mutexattr_init(&mtconf); Assert(err == 0);
+        err = pthread_mutexattr_settype(&mtconf, PTHREAD_MUTEX_RECURSIVE); Assert(err == 0);
+    }
+
+    pthread_mutex_t *mutex = (pthread_mutex_t*)this->mutexPtr;
```

Comment:
Is this properly defined behavior? If the `this->mutexPtr` is actually a pointer type, why not declare it as a `void *` in the header if you want it to be opaque, and then cast the pointer here? That should be well defined behavior as far as I'm aware.

Reply:
‘void*’ requires an additional allocation

Reply:
Oh, I think I've misunderstood this code then. I thought you were treating the `mutexPtr` as a pointer, but you are actually using the array-interpreted-as-pointer-to-itself as the mutex pointer?

I'm not sure what the benefit of this is over just having the `mutexPtr` in the class be defined as a `pthread_mutex_t` directly. Is this because you don't want to include the pthread header in the header for this class?

Reply:
Yes. PAL also uses it and including pthread header on each consumer complicates the things